Shopify/ruby-lsp
# Recover from files with initial syntax errors

### Motivation

We are getting some errors on user's machines of the `document` instance being `nil`. The only possibility for this, is if we weren't able to parse the first version of the file due to syntax errors and never recovered.

### Implementation

This PR attempts to make this process more robust by always inserting the document into the store regardless of syntax errors. If a syntax error occurs, the `tree` will be `nil`, which we have to handle in requests.

When edits are made to the file with syntax errors, `push_edits` will take care of re-parsing.

### Automated Tests

Added a couple of examples and changed some existing ones.

### Manual Tests

If you wanna try this manually
1. Start the LSP on this branch
2. Created a new file and save it with a syntax error (e.g.: `def Foo`)
3. Close the file to clear its state from the store
4. Re-open the file (this re-open will fail to parse)
5. Fix the syntax error by adding `end`
6. Verify that the LSP recovered and features are now available
